lib/http-client.ts

```diff
@@ -68,6 +114,92 @@ export class HttpClient {
         return this.send('DELETE', url, null, headers);
     }
 
+    /**
+     * 
+     * Method used to generate predefined reusable methods for given set of endpoints.
+     * 
+     * Example:
+     * 
+     * let endpoints =  { 
+     *      getPosts : {
+     *          path : "/posts",
+     *          "methods" : "GET"
+     *      }
+     *  }
+     * 
+     * Below method will create a dynamic function in HttpClient prototype with name "getPosts" 
+     * and the method can be directly called from test like `http.getPosts()`.
+     * 
+     * @param endpoints
+     * @returns {HttpClient}
+     */
+    
+    registerEndpoints(endpoints:IendPointTemplate):HttpClient {
+        let self:HttpClient = this;
+        Object.keys(endpoints).forEach(function (endPointName:string) {
+            var endPointDetails = endpoints[endPointName];
+            HttpClient.prototype[endPointName] = function (parameters?:any, body?:any, headers?:any):ResponsePromise {
+                //parse wildcard routes before making API call.
+                let parsedUrl = HttpClient.parseWildCardRoutes(endPointDetails.path, parameters || {});
+                return self.send(endPointDetails.method, parsedUrl, body, endPointDetails.headers || headers);
+            }
+        });
+        return this;
+    }
+
+    /**
+     * 
+     * Method to convert wildcard routes with their respective values.
+     * 
+     * Example 1:
+     * let path = "/products/{productId}";
+     * let wildcardParam = { productId : 1 };
+     * 
+     * OUTPUT: "/products/1"
+     * 
+     * Example 2:
+     * 
+     * If the path doesn't have any wildcard pattern, then this method will append parameter as url-encoded-form.
+     * 
+     * let path = "/products";
+     * let wildcardParam = { productId : 1 };
+     * 
+     * OUTPUT: "/products?productId=1"
+     * 
+     * @param path
+     * @param wildcardParam
+     * @returns {string}
+     */
+    private static parseWildCardRoutes(path:string, wildcardParam:any):string {
+        var regexObject,isFirstUrlParam = true;
+        for (var param in wildcardParam) {
+            if (wildcardParam.hasOwnProperty(param) && path.indexOf("{" + param + "}") >= 0) {
+                regexObject = new RegExp("{" + param + "}", "g");
+                path = path.replace(regexObject, HttpClient.convertParamsToStringObject(wildcardParam[param]));
+            } else {
+                if(isFirstUrlParam) {
```

Comment:
personally, I would advise to avoid over-complicating this logic, that as far as I understood, automatically adds unknown parameters into the URL as request parameters. I would just write the endpoints like so
```
/my/endpoint/{id}?requestParam1={paramValue1}&param2={value2}
```

Reply:
I agree.But consider the below API endpoint example
`
/products will return all products.
/products?page=1 will return all products from page 1
/products?ordered_by=sudharsan will return all products ordered by sudharsan
/products?ordered_by=sudharsan&page=1 will return products from page 1 ordered by sudharsan
`
So to handle all above use cases, i thought we need to support dynamic parameters for every endpoints.